### PR TITLE
add concretizer mode to ignore versions that are moving targets, use that in CI

### PR DESCRIPTION
**Pull request by haampie**
Add a concretizer mode that restricts the concretizer to checksummed versions.

Checksummed versions are currently either:

1. tarballs with any supported shasum
2. git commits with 40 char shasum

It also enables this in Gitlab CI to avoid building moving targets.

The mode is enabled by setting the undocumented environment variable

```SPACK_CONCRETIZER_REQUIRE_CHECKSUM```

and we can decide on nice config option for this later.

This PR depends the following for it to pass in Gitlab CI:
- [x] #39644 
- [x] #38738 
- [x] #39635
- [x] #39637

### CODE REVIEW

**Comment by haampie** [comment]
@spackbot run pipeline

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -569,6 +570,46 @@ def keyfn(x):
...
+    # alternative is to instantiate a fetch strategy object, which is a lot of
+    # work for a simple filter. Also semantics are not always the same in different
+    # use cases, e.g. "cacheable" doesn't necessarily imply "checksummable".
+    version, option_dict = version_with_options
+    if isinstance(version, spack.version.StandardVersion):
+        if any(hash_type in option_dict for hash_type in spack.util.crypto.hashes.keys()):
+            return True
+        if "commit" in option_dict:
+            # Apparently we allow short git commit shas... require 40 chars.
+            return len(option_dict["commit"]) == 40
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: Short git commits can't give you the wrong source later. If the commit becomes ambiguous you get an error.

File: etc/spack/defaults/concretizer.yaml [hunk 2]
```diff
@@ -41,4 +41,7 @@ concretizer:
     # "none": allows a single node for any package in the DAG.
     # "minimal": allows the duplication of 'build-tools' nodes only (e.g. py-setuptools, cmake etc.)
     # "full" (experimental): allows separation of the entire build-tool stack (e.g. the entire "cmake" subDAG)
-    strategy: none
\ No newline at end of file
+    strategy: none
+  # When set to true, the concretizer only considers checksummed versions of packages. This
+  # includes checksummed tarballs and long commit shas. Versions based on git branches are ignored.
+  checksummed: false
```
Comment (becker33) [inline etc/spack/defaults/concretizer.yaml]: I think this should go under a `versions` heading, because there are other things related to versioning that could go in the concretizer config. The `deprecated` config option should probably move here from the `config` section.

I would think the default config should be:

```
concretizer:
  versions:
    checksummed: false
    deprecated: false
```

**Review by becker33 — DISMISSED** [review]
I think a better name would be `concretizer:versions:reproducible_only` since git commits don't come with a checksum in Spack (although I recognize the term is accurate since they are checksummed by git).

File: lib/spack/spack/solver/asp.py (hunk 1 above)
Comment (haampie) [reply lib/spack/spack/solver/asp.py]: They can if the repo gets compromised, or with man in the middle attacks. Generating commit shas with a given prefix is trivial, see for example the git history in this fun repo https://github.com/dj-shin/gitchain ;)

File: etc/spack/defaults/concretizer.yaml (hunk 2 above)
Comment (haampie) [reply etc/spack/defaults/concretizer.yaml]: `config:deprecated` also toggles a prompt for "fetch anyway?" on `spack stage/install` of concrete specs/envs. So it doesn't feel like it fits under `concretizer:deprecated`.

**Comment by haampie** [comment]
Sigh. `+1,301 -571` thanks to black formatting :(

@tgamblin can we increase the black style line length for package.py?

**Comment by haampie** [comment]
@alecbcs maybe you can incorporate [some of this](https://gist.github.com/haampie/3441e8608409732a6d3cb7841f524aab) in your version check scripts

File: etc/spack/defaults/concretizer.yaml (hunk 2 above)
Comment (haampie) [reply etc/spack/defaults/concretizer.yaml]: regarding naming I also think `concretizer:checksummed:true` makes sense, but `concretizer:checksummed:false` sounds ambiguous. `require_checksum:true|false` sounds more clear, but is rather long

**Comment by haampie** [comment]
> concretizer:versions:reproducible_only

I think that's also ok, but

> since git commits don't come with a checksum in Spack

now they do

**Comment by haampie** [comment]
@scheibelp can you briefly look at this commit 4ab7a5820bb4fdbbc5f8e6eca5e9b269c468f3bd since it touches code you wrote

I don't think the previous implementation made a lot of sense, although I'm to blame for not taking the time to review it properly ahead of ISC / 0.20 release.

What I think it did:

1. goes over all requirements
2. extracts the specs
3. takes the versions, distinguishes between:
   1. abstract, if so, collect all possible versions matching abstract
   2. otherwise collect the specific version
4. creates `Spec` objects of them (?!)
5. passes those to some function that declares new versions

There's a ton of things that don't make sense there:

1. It would define versions for specs that are statically outside the possible dep graph
2. For some reason the package version definitions were pulled in in (3.i), even though those were all _just_  defined/declared in possible_versions[pkg_name] since they're from package.py
3. For requirements like `+a` (not mentioning a version at all) the 3.i case kicks in and you basically copy *all* possible versions `[1.0, 1.1, 2.0, ...]` => instantiate a new `Spec` object for *all* of them `[pkg@=1.0, pkg@=1.1, pkg@=2.0, ...]` => send them to some function that declares versions from specs normally used for cli/dev specs => which extracts them to versions `[1.0, 1.1, 2.0, ...]` again and tests whether they are already in `possible_versions`, which *all of them are by construction*, and ... the function returns.

So, what I've changed is: only concrete versions in requirements are declared as extra possible versions (unless the new config option in this PR is toggled on)

File: etc/spack/defaults/concretizer.yaml (hunk 2 above)
Comment (tgamblin) [reply etc/spack/defaults/concretizer.yaml]: Should we call these versions "safe" like we do in `spack info`?  I don't particularly like the term so I'm not sure I am really advocating for it, but I think we should be consistent.  Not opposed to changing the language in `spack info` as part of this PR to match what goes in config.

I think the allowed version types should be in a list, e.g.:

```yaml
concretizer:
    versions:
        include: [safe, preferred]
        exclude: [deprecated]
```

and maybe `config:deprecated:true` should be phased out for `concretizer:versions:include:[deprecated]`.  It should probably be synonymous with that for some fairly large number of releases.

Thoughts?

**Comment by scheibelp** [comment]
> What I think it did:

Your description looks right

> creates Spec objects of them (?!)

Yes, this comes from using the already-existing `add_concrete_versions_from_specs`: you have to track name and version. And that's how the other version definition components (which predate https://github.com/spack/spack/pull/37687) do it. 

> For some reason the package version definitions were pulled in in (3.i), even though those were all just defined/declared in possible_versions[pkg_name] since they're from package.py

As you point out in

`# TODO: stop re-defining existing versions, just relabel/adjust weights.`

The versions are re-added to promote their new weight (otherwise they are not preferred). One could complain about this but I'll note this behavior was not added by https://github.com/spack/spack/pull/37687 - the PR simply continued the existing strategy.

> It would define versions for specs that are statically outside the possible dep graph

You mean in the sense that it produces ASP definitions for all requirements regardless of whether they concern a possible dependency? True enough.

> For requirements like +a (not mentioning a version at all) the 3.i case kicks in and you basically copy all possible versions

I definitely screwed up there. It should only do it if `spec.versions` is defined.

> So, what I've changed is: only concrete versions in requirements are declared as extra possible versions (unless the new config option in this PR is toggled on)

That sounds like an alteration of the behavior described at https://github.com/spack/spack/pull/36650#issue-1654773724 which I believe I derived from a discussion with the full team.

**Comment by haampie** [comment]
> The versions are re-added to promote their new weight (otherwise they are not preferred). One could complain about this but I'll note this behavior was not added by https://github.com/spack/spack/pull/37687 - the PR simply continued the existing strategy.

For preferences I guess the current implementation is OK. I just added a note for future refactoring, because it should not be necessary to duplicate version declarations, when they could instead be marked or "promoted" as preferred [or maybe preferences could even just enter the default "concretization order" imposed on the set of versions, so we order by `(preferred in packages.yaml, preferred in package.py, deprecated or not, finite or not, version)` tuples.]

Regarding **requirements** though, only concrete versions can cause *new* declarations, like when you do `packages:zlib:require: @git.develop=1.3.1`, or `packages:pkg:require ^zlib@=1.3.1` assuming the git version `git.develop=1.3.0-rc1` and standard version `=1.3.1` are not defined in package.py / externals. That's the rare / exceptional case.

For the common case of non-concrete version requirements, it's just a restriction on the solution space for versions, there's no need to declare anything. In particular this dance from abstract version -> list of matching concrete possible versions -> list of specs is completely redundant, because all of these are dropped as they already exist.

So, (a) there is no need to figure out what exact versions a requirement `zlib@1` can satisfy, in principle the solver does that anyways (b) but to be nice, we can just check that there's _at least one_ possible version that can satisfy the requirement -- if not, we raise an error that informs the user that their config cannot give any solutions.

In short: no need to collect / construct all possible matching concrete versions, just figuring out one match is sufficient to be able to avoid raising a ConfigError. That's what this commit does for abstract version requirements.

File: etc/spack/defaults/concretizer.yaml (hunk 2 above)
Comment (haampie) [reply etc/spack/defaults/concretizer.yaml]: That suggests you can do `exclude: [safe]` which is nonsensical :p

Comment (haampie) [reply etc/spack/defaults/concretizer.yaml]: We can also fix it with an undocumented environment variable, and decide on a public facing API later. Cause I'd like to use this to fix the current broken CI on develop.

**Comment by scheibelp** [comment]
> For the common case of non-concrete version requirements, it's just a restriction on the solution space for versions, there's no need to declare anything.

OK I think that makes sense (i.e. I see I was adding something redundant):

* The requirement ordering is established by a separate weighting system (`requirement_weight` in `concretize.lp`)
  * And `test/concretize_requirements.py:test_oneof_ordering` checks this
* `_spec_clauses -> spec_versions` updates `version_constraints`, which handles association of abstract version ranges with concrete ones (i.e. making sure the concretizer knows that `1:1.5` is satisfied by `1.1`) so we don't need `add_concrete_versions_from_specs` for that
  * And `test/concretize_requirements.py:test_require_truncated` checks this

File: etc/spack/defaults/concretizer.yaml (hunk 2 above)
Comment (haampie) [reply etc/spack/defaults/concretizer.yaml]: This PR is no longer about finding a name for a config option, so hitting resolve here

File: lib/spack/spack/test/concretize.py [hunk 3]
```diff
@@ -2208,3 +2216,39 @@ def test_solution_without_cycles(self):
...
+        # git branch is a moving target
+        ("1.2.3", {"branch": "releases/1.2"}, False),
+        # git ref is a moving target
+        ("git.branch=1.2.3", {}, False),
+    ],
+)
+@pytest.mark.only_clingo("Use case not supported by the original concretizer")
+def test_drop_moving_targets(v_str, v_opts, checksummed):
+    import spack.solver.asp
+
```
Comment (alalazo) [inline lib/spack/spack/test/concretize.py]: Can we move import at the top?

File: lib/spack/spack/test/concretize.py [hunk 4]
```diff
@@ -2208,3 +2216,39 @@ def test_solution_without_cycles(self):
...
+        ("1.2.3", {"sha384": f"{1:096x}"}, True),
+        ("1.2.3", {"sha512": f"{1:0128x}"}, True),
+        # no digest key
+        ("1.2.3", {"bogus": f"{1:064x}"}, False),
+        # git version with full commit sha
+        (f"{1:040x}=1.2.3", {}, True),
+        # git version with short commit sha
+        (f"{1:07x}=1.2.3", {}, False),
+        # git tag is a moving target
+        ("1.2.3", {"tag": "v1.2.3"}, False),
```
Comment (alalazo) [inline lib/spack/spack/test/concretize.py:2245]: Can we unit test `tag` + `commit` too?

File: lib/spack/spack/solver/asp.py [hunk 5]
```diff
@@ -1887,54 +1918,38 @@ def key_fn(item):
...
                 else:
-                    satisfying_versions = self._check_for_defined_matching_versions(pkg_class, v)
-                    # Amongst all defined versions satisfying this specific
-                    # preference, the highest-numbered version is the
-                    # most-preferred: therefore sort satisfying versions
-                    # from greatest to least
-                    version_defs.extend(sorted(satisfying_versions, reverse=True))
+                    # TODO: stop re-defining existing versions, just relabel/adjust weights.
+                    matches = [x for x in self.possible_versions[pkg_name] if x.satisfies(v)]
+                    # TODO: use concretization order instead of version order
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]: Not clear what you mean by these comments.

File: lib/spack/spack/solver/asp.py [hunk 6]
```diff
@@ -1887,54 +1918,38 @@ def key_fn(item):
...
-                "Preference for version {0} does not match any version"
-                " defined for {1} (in its package.py or any external)".format(str(v), pkg_name)
-            )
-        return satisfying_versions
-
-    def add_concrete_versions_from_specs(self, specs, origin):
+    def define_ad_hoc_versions_from_specs(self, specs, origin, require_checksum: bool):
         """Add concrete versions to possible versions from lists of CLI/dev specs."""
-        for s in spack.traverse.traverse_nodes(specs):
+        for s in traverse.traverse_nodes(specs):
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:1954]: (nitpick) Why are we importing this module using `from` (and all the rest using fully qualified names?) :thinking:

File: lib/spack/spack/solver/asp.py [hunk 7]
```diff
@@ -1943,9 +1958,11 @@ def add_concrete_versions_from_specs(self, specs, origin):
             if version is None or any(v == version for v in self.possible_versions[s.name]):
                 continue
 
-            self.declared_versions[s.name].append(
-                DeclaredVersion(version=version, idx=0, origin=origin)
-            )
+            if require_checksum and not (isinstance(version, vn.GitVersion) and version.is_commit):
+                continue
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]: Here we have the opportunity to collect errors and raise early during setup, instead of waiting for clingo to have `errors` in the answer set.

File: lib/spack/spack/solver/asp.py [hunk 8]
```diff
@@ -2447,78 +2463,67 @@ def literal_specs(self, specs):
...
+                    continue
+
+                s.attach_git_version_lookup()
+                v = versions.concrete
+
+                if not v:
+                    # If the version is not concrete, check it's statically concretizable. If
+                    # not throw an error, which is just so that users know they need to change
+                    # their config, instead of getting a hard to decipher concretization error.
+                    if not any(x for x in self.possible_versions[name] if x.satisfies(versions)):
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:2493]: We should leave a note in the docstring that now this function assumes that the possible versions from package.py are defined already.

File: lib/spack/spack/solver/asp.py [hunk 9]
```diff
@@ -2447,78 +2463,67 @@ def literal_specs(self, specs):
...
+                            f"cannot match any known version from package.py or externals"
+                        )
+                    continue
+
+                if v in self.possible_versions[name]:
+                    continue
+
+                # If concrete an not yet defined, conditionally define it, like we do for specs
+                # from the command line.
+                if not require_checksum or (isinstance(v, vn.GitVersion) and v.is_commit):
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]: Modulo de Morgan law, can we collect this in a function to be used both here and for cli specs?

**Review by alalazo — CHANGES_REQUESTED** [review]
A few minor questions / requests

File: lib/spack/spack/solver/asp.py (hunk 6 above)
Comment (haampie) [reply lib/spack/spack/solver/asp.py:1954]: I guess it's like the spack.version as vn convention elsewhere.

File: lib/spack/spack/test/concretize.py (hunk 3 above)
Comment (haampie) [reply lib/spack/spack/test/concretize.py]: It's done like this in 6 other tests so I assumed there was a reason for that.

File: lib/spack/spack/solver/asp.py (hunk 5 above)
Comment (haampie) [reply lib/spack/spack/solver/asp.py]: Let me remove the comment since it's irrelevant to the PR. I've explained it in a comment above: there's no need to duplicate versions if you can just mark them, or include preference criterion in the sort key. The second comment is about when a preference matches multiple existing versions, they should not be ordered by version, but by concretization order `(preferred in package.py, not deprecated, not finite, version)`.

File: lib/spack/spack/test/concretize.py (hunk 3 above)
Comment (haampie) [reply lib/spack/spack/test/concretize.py]: narrator: there was not

**Review by alalazo — APPROVED** [review]
